zebrad/src/components/sync/recent_sync_lengths.rs

```diff
@@ -0,0 +1,87 @@
+//! A channel which holds a list of recent syncer response lengths.
+
+use tokio::sync::watch;
+
+#[cfg(test)]
+mod tests;
+
+/// A helper type which holds a list of recent syncer response lengths.
+/// These sync lengths can be used to work out if Zebra has reached the end of the chain.
+///
+/// New lengths are added to the front of the list.
+/// Old lengths are dropped if the list is longer than `MAX_RECENT_LENGTHS`.
+//
+// TODO: disable the mempool if obtain or extend tips return errors?
+#[derive(Debug)]
+pub struct RecentSyncLengths {
+    ///
+    sender: watch::Sender<Vec<usize>>,
+
+    // TODO: Replace with calls to `watch::Sender::borrow` once Tokio is updated to 1.0.0 (#2573)
+    recent_lengths: Vec<usize>,
+}
+
+impl RecentSyncLengths {
+    /// The maximum number of lengths sent by `RecentSyncLengths`.
+    ///
+    /// Older lengths are dropped.
+    ///
+    /// This length was chosen as a tradeoff between:
+    /// * clearing temporary errors and temporary syncs quickly
+    /// * distinguishing between temporary and sustained syncs/errors
+    /// * activating the syncer shortly after reaching the chain tip
+    pub const MAX_RECENT_LENGTHS: usize = 4;
+
+    /// Create a new instance of [`RecentSyncLengths`]
+    /// and a [`watch::Receiver`] endpoint for receiving recent sync lengths.
+    pub fn new() -> (Self, watch::Receiver<Vec<usize>>) {
+        let (sender, receiver) = watch::channel(Vec::with_capacity(Self::MAX_RECENT_LENGTHS));
```

Comment:
Minor: I think it doesn't make sense to preallocate here for the initial value. It will stay empty and will be replaced by the next call to `sender.send`.

```suggestion
        let (sender, receiver) = watch::channel(Vec::new());
```

Reply:
You're right, we'd need to borrow the vector from the sender for that to work.

Fixed in 55318413.